[MLv2] Add functions to support viz settings in FE

bshepherdson opened the pull request:
Viz settings contain legacy MLv1 field refs for the selected columns of
a query. We need to round-trip those settings into MLv2 `:fields` lists
and back to legacy refs. 

The table column picker in the viz settings is migrated to MLv2.

This adds:
- `visible-columns` to get the full set of columns in scope on a query
  - this includes `:selected?` information for the columns
- `legacy-field-ref` to convert an MLv2 field ref to legacy MLv1
- `find-column-index-from-legacy-ref` to match a legacy ref against a
  list of MLv2 columns

<img width="553" alt="Screenshot 2023-08-14 at 18 19 15" src="https://github.com/metabase/metabase/assets/8542534/9a863e05-0c2e-48cd-ab6a-d7e35186cf8f">

bshepherdson committed:
[MLv2] Add functions to support viz settings in FE

Viz settings contain legacy MLv1 field refs for the selected columns of
a query. We need to round-trip those settings into MLv2 `:fields` lists
and back to legacy refs.

This adds:
- `visible-columns` to get the full set of columns in scope on a query
  - this includes `:selected?` information for the columns
- `legacy-field-ref` to convert an MLv2 field ref to legacy MLv1
- `find-column-index-from-legacy-ref` to match a legacy ref against a
  list of MLv2 columns

It also makes lib.field/add-field and remove-field work properly for
custom expression columns rather than being a special case.

ranquild commented on frontend/src/metabase-lib/comparison.ts:
```diff
@@ -8,3 +9,17 @@ export function areLegacyQueriesEqual(
+export function findColumnIndexForColumnSetting(
+  query: Query,
+  stageIndex: number,
+  columns: ColumnMetadata[],
```
```suggestion
  columns: (ColumnMetadata | DatasetColumn)[],
```

metamben replied:
In fact, it seems I implemented this suggestion...

ranquild commented on frontend/src/metabase-lib/comparison.ts:
```diff
@@ -1,5 +1,6 @@
 import * as ML from "cljs/metabase.lib.js";
-import type { DatasetQuery } from "metabase-types/api";
+import type { DatasetQuery, FieldReference } from "metabase-types/api";
```
```suggestion
import type { DatasetColumn, DatasetQuery, FieldReference } from "metabase-types/api";
```

metamben commented on src/metabase/lib/js.cljs:
````diff
@@ -418,6 +421,67 @@
+  which defines the set of columns to return. All the options default to *true*, ie. everything is returned by default.
+  ```
+  {
+    include_joined: true,
+    include_expressions true,
+    include_implicitly_joinable: true,
````
Shouldn't these be in camelCase?

ranquild replied:
btw @bshepherdson the FE don't use these options anymore, could we drop them entirely? so it would work like all of them are true

metamben commented on src/metabase/lib/js.cljs:
````diff
@@ -418,6 +421,67 @@
+  }
+  ```"
+  [a-query stage-number ^js js-options]
+  (let [options     (merge visible-columns-defaults
+                           (when-let [opts (some-> js-options js-keys set)]
+                             (cond-> {}
+                               (opts "includeJoined")
+                               (assoc :include-joined? (.-includeJoined js-options))
+                               (opts "includeExpressions")
+                               (assoc :include-expressions? (.-includeExpressions? js-options))
+                               (opts "includeImplicitlyJoinable")
+                               (assoc :include-implicitly-joinable? (.-includeImplicitlyJoinable js-options)))))
````
```suggestion
  (let [options     (cond-> visible-columns-defaults
                      (gobj/containsKey js-options "includeJoined")
                      (assoc :include-joined? (.-includeJoined js-options))
                      (gobj/containsKey js-options "includeExpressions")
                      (assoc :include-expressions? (.-includeExpressions? js-options))
                      (gobj/containsKey js-options "includeImplicitlyJoinable")
                      (assoc :include-implicitly-joinable? (.-includeImplicitlyJoinable js-options)))
```
Or maybe even just use `(gobj/get js-options "includeJoined" true)` instead of relying on `visible-columns-defaults`.

bshepherdson replied:
I went with the latter option, that works fairly well.

metamben commented on src/metabase/lib/js.cljs:
```diff
@@ -418,6 +421,67 @@
+
+(defn ^:export find-column-index-from-legacy-ref
+  "Given a list of columns (either JS `data.cols` or MLv2 `ColumnMetadata`), search through it for a field matching the
+  given legacy field ref. Returns the index, or -1 if no matching column is found."
```
Returning -1 is quite standard (especially in Java and C), but I wonder if returning `nil` would be more comfortable. (The implementation of the function would definitely be a bit more succinct.)

bshepherdson replied:
It's also conventional in Clojure to return nil rather than empty collections, and to return Clojure collections rather than JS arrays, and we do both those things in this file. This is the JS shim file, and I argue for following JS conventions in here.

metamben replied:
If JS people expect -1 for not found from such a function, then by all means keep it. 👍

metamben commented on src/metabase/lib/js/metadata.cljs:
```diff
@@ -24,7 +25,7 @@
 
 (defn- object-get [obj k]
-  (when obj
+  (when (and obj (js-in k obj))
```
```suggestion
  (when (string? k)
```

bshepherdson replied:
What? Did the suggestion get mangled?

metamben replied:
I really meant to use `(string? k)` as the test. Wouldn't that be simpler? (The following line Just Works with `obj` being `nil` and when `k` not being a property of `obj`. I think the only difference is, it doesn't blow up when `k` is not a string. Is that undesirable?

bshepherdson replied:
Ah, I see. I think back when I made these changes I was trying to debug what turns out to be the `node-repl` being bad. `goog.object` doesn't work properly there, and was throwing in cases where the browser returns `nil` no problem. I think `(string? k)` suffices here, yes.

metamben commented on src/metabase/lib/js/metadata.cljs:
```diff
@@ -82,13 +83,27 @@
+  If this function returns nil for a given key, the original key is preserved.
+  Use [[excluded-keys]] to drop keys from the input.
+
+  Defaults to nil, which means no renaming is done."
+  identity)
+
+(defmethod rename-key-fn :default [_]
+  nil)
```
Wouldn't it be simpler if it defaulted to `identity`? That seems to be the straightforward way to say "no change".

bshepherdson replied:
Doing that makes the renaming functions harder to write. With this approach they can be maps called like functions, and things that don't appear in the map are preserved.

metamben commented on src/metabase/lib/js/metadata.cljs:
```diff
@@ -365,15 +385,15 @@
   (some-> metadata :segments deref (get segment-id) deref))
 
 (defn- tables [metadata database-id]
-  (for [[_id table-delay] (some-> metadata :tables deref)
-        :let              [a-table (some-> table-delay deref)]
-        :when             (and a-table (= (:db-id a-table) database-id))]
+  (for [[_id table-delay]  (some-> metadata :tables deref)
+        :let               [a-table (some-> table-delay deref)]
+        :when              (and a-table (= (:db-id a-table) database-id))]
     a-table))
 
 (defn- fields [metadata table-id]
-  (for [[_id field-delay] (some-> metadata :fields deref)
-        :let              [a-field (some-> field-delay deref)]
-        :when             (and a-field (= (:table-id a-field) table-id))]
+  (for [[_id field-delay]  (some-> metadata :fields deref)
+        :let               [a-field (some-> field-delay deref)]
+        :when              (and a-field (= (:table-id a-field) table-id))]
```
Are these changes needed?

bshepherdson replied:
No, I'm not sure where those came from. Fixed.

metamben reviewed (approved):
I have only some nits.

ranquild committed:
Port changing query columns to MLv2

Part of #32501

metamben committed:
[MLv2] Fix some BE and e2e tests broken by porting viz-settings

Part of #32501

Allows removal of implicitly joinables w/o explicit fields clause.

As demonstrated by metabase#32373 in table-column-settings.cy.spec.js,
fields coming from source questions can be implicitly joinable according
to :lib/source and still appear without an explicit :fields clause.

metamben commented on src/metabase/lib/metadata/calculation.cljc:
```diff
@@ -451,17 +451,19 @@
+    [:include-joined?                              {:optional true} :boolean]
+    [:include-expressions?                         {:optional true} :boolean]
+    [:include-implicitly-joinable?                 {:optional true} :boolean]
+    [:include-implicitly-joinable-for-source-card? {:optional true} :boolean]]])
```
I think this flag points to some inconsistency in the product. Apparently, in some places (e.g., viz settings) we want to see implicitly joinable fields for _source cards_, in other places (orderable and breakoutable columns) we don't.

metamben replied:
@bshepherdson see @luizarakaki's  [comment](https://github.com/metabase/metabase/issues/30950#issuecomment-1567139855) related to this.

metamben commented on test/metabase/lib/card_test.cljc:
```diff
@@ -78,16 +78,3 @@
-      (is (= ["ID" "SUBTOTAL" "TOTAL" "TAX" "DISCOUNT" "QUANTITY" "CREATED_AT" "PRODUCT_ID" "USER_ID"]
-             (mapv :name (get-in lib.tu/mock-cards [:orders :result-metadata]))))
-      (is (= ["ID" "SUBTOTAL" "TOTAL" "TAX" "DISCOUNT" "QUANTITY" "CREATED_AT" "PRODUCT_ID" "USER_ID"]
-             (mapv :name (lib/visible-columns venues-query)))))))
```
@camsaul Removing this test is questionable, but required for this PR to work.

camsaul replied:
instead of removing this, can't you call this with the `include-implicitly-joinable-source-card` setting disabled?

metamben replied:
That can be done, if we don't really mean what the test was saying.

metamben commented on test/metabase/lib/stage_test.cljc:
```diff
@@ -178,7 +178,22 @@
+                  :base-type                :type/DateTime
+                  :lib/source               :source/implicitly-joinable
+                  :lib/desired-column-alias "USERS__via__USER_ID__LAST_LOGIN"}]
                 (lib/visible-columns query)))))))
```
@camsaul Reverting this test is questionable, but required for this PR to work. In this case at least I know that issue I fixed in #30950 by removing implicitly joinable fields everywhere is fixed in this PR differently (i.e., only for orderable and breakoutable columns).

camsaul replied:
Can't you just update the call to `visible-columns` to include the new option to exclude implicitly joinable columns for source Cards? Actually it would be even better if you tested both scenarios (with the new option enabled = include the implicitly joinable columns; with the new option disabled = exclude the columns)

metamben replied:
That can be done too.

alxnddr commented:
I found a few cases that do not work as expected, please have a look:

1) Nested question
- Create a stuctured question Q1 from the Orders table
- New -> Question -> Saved Questions -> Q1
- In the Notebook mode untick the "Tax" column
- Click "Visualize"
- Try add the "Tax" column in the sidebar

Actual:
The column is added to the table but not removed from the More column section.

2) A saved question joined with a table
- Create a stuctured question Q1 from the Orders table
- New -> Question -> Saved Questions -> Q1
- Join with products 
- Click "Visualize"
- In the sidebar try adding the "ID" column from the "User" table

Actual:
An "ID" table is added to the table but it is a duplicate of the Orders.ID. The column has not been removed from the sidebar.

3) Tree-like self-join
- New -> Question -> Orders
- Select Id and Tax columns only
- Join Orders on Id = UserId
- For the joined table select UserId and Tax columns only
- Click "Visualize"
- In the sidebar try removing "Tax" column

Actual:
After the query reruns, the column "Orders -> Tax" is removed too.

camsaul commented on src/metabase/lib/convert.cljc:
```diff
@@ -155,20 +174,17 @@
-                       (update stage k ->pMBQL)))
-                   stage
-                   (disj stage-keys :aggregation :expressions))]
+                    (fn [stage k]
```
old indentation was correct 🙃

ranquild commented:
> 1) Nested question
> - Create a stuctured question Q1 from the Orders table
> - New -> Question -> Saved Questions -> Q1
> - In the Notebook mode untick the "Tax" column
> - Click "Visualize"
> - Try add the "Tax" column in the sidebar
> 
> Actual:
> The column is added to the table but not removed from the More column section.

The series of the calls is the following:
- When adding a new column (Tax from the above) in the `table.columns` setting, the FE calls `legacyFieldRef`, which returns `["field", "TAX", {"base-type": "type/Float"}]`
- Query is re-run. `/api/dataset` returns `data.cols` with new columns, and for Tax `field_ref` is `['field', 11, null]`
- The FE calls `findColumnIndexesFromLegacyRefs`, passing `data.cols` and field refs from `table.columns`:
```
const datasetIndexes = Lib.findColumnIndexesFromLegacyRefs(
    query,
    STAGE_INDEX,
    datasetColumns,
    fieldRefs,
  );
```

For Tax column the returned index is `-1`, and this makes the FE think that the column is not included in the response.

Expected behavior:
- The correct index should be returned

ranquild commented:
> 2) A saved question joined with a table
> - Create a stuctured question Q1 from the Orders table
> - New -> Question -> Saved Questions -> Q1
> - Join with products 
> - Click "Visualize"
> - In the sidebar try adding the "ID" column from the "User" table
> 
> Actual:
> An "ID" table is added to the table but it is a duplicate of the Orders.ID. The column has not been removed from the sidebar.

In this case the FE calls `addField` on a column obtained from `visibleColumns` but it has no effect. The query isn't changed.

ranquild commented:
> 3) Tree-like self-join
> - New -> Question -> Orders
> - Select Id and Tax columns only
> - Join Orders on Id = UserId
> - For the joined table select UserId and Tax columns only
> - Click "Visualize"
> - In the sidebar try removing "Tax" column
> 
> Actual:
> After the query reruns, the column "Orders -> Tax" is removed too.

I managed to reproduce it by removed `Orders -> Tax` and rerunning the query. After that the original `Tax` column coming from the source table disappeared from the settings.

In this case we have the following fields refs in column settings
<img width="508" alt="Screenshot 2023-08-17 at 13 31 10" src="https://github.com/metabase/metabase/assets/8542534/b2bbdddc-f9b0-4ed6-adf2-9c3dac011390">

For some reason `findColumnIndexesFromLegacyRefs` fails to find the correct column for just `Tax` (notice it doesn't have `join-alias` in the image above) but assigns the index for the incorrect field ref with `join-alias`.

`data.cols` here don't have the `join-alias` as expected:

<img width="518" alt="Screenshot 2023-08-17 at 13 37 39" src="https://github.com/metabase/metabase/assets/8542534/9d5b7ea0-c758-4670-b8ee-cca5a8732e4e">

This seems to be another issue with how `findColumnIndexesFromLegacyRefs` finds columns.

ranquild committed:
Add e2e repro cases for three nested query issues

These are discussed by @alxnddr here:
https://github.com/metabase/metabase/pull/32698#issuecomment-1681384427

camsaul commented on src/metabase/lib/join/util.cljc:
```diff
@@ -0,0 +1,89 @@
+            source-field-id-name (:name (lib.metadata/field query fk-field-id))]
+        (format-implicit-join-name table-name source-field-id-name)))))
+
+(mu/defn desired-alias :- ::lib.schema.common/non-blank-string
+  "Desired alias for a Field e.g.
+
+    my_field
+
+    OR
+
+    MyJoin__my_field
+
+  You should pass the results thru a unique name function."
+  [query          :- ::lib.schema/query
+   field-metadata :- lib.metadata/ColumnMetadata]
+  (if-let [join-alias (or (current-join-alias field-metadata)
+                          (implicit-join-name query field-metadata))]
+    (joined-field-desired-alias join-alias (:name field-metadata))
+    (:name field-metadata)))
```
I don't really agree with moving this here since even a Field without a join alias still has a desired alias

bshepherdson replied:
I'd support an alternative home, it just needed out of `metabase.lib.field` to avoid a circular dep.

`metabase.lib.field.util`?

camsaul commented on test/metabase/lib/js_test.cljs:
```diff
@@ -74,3 +74,7 @@
               {:lib/type :option/join.strategy, :strategy :inner-join}]
              (vec strategies))))))
+
+(deftest ^:parallel find-column-index-from-legacy-ref-test
+
+ )
```
????

bshepherdson replied:
Leftovers from development that got accidentally committed to the working branch as I was prepping to leave for vacation. There should be some tests written for the custom logic in that CLJS function since it's not a trivial wrapper of another MLv2 function.

camsaul commented on src/metabase/lib/js.cljs:
```diff
@@ -427,6 +431,153 @@
+        (lib.equality/find-closest-matches-for-refs
+         a-query
+         stage-number
+         field-refs
+         (mapv key new-columns->columns)
+         {:keep-join? true})]
+    (update-keys matches new-columns->columns)))
+
+(defn ^:export find-column-indexes-from-legacy-refs
+  "Given a list of columns (either JS `data.cols` or MLv2 `ColumnMetadata`) and a list of legacy refs, find each ref's
+  corresponding index into the list of columns.
+
+  Returns a parallel list to the refs, with the corresponding index, or -1 if no matching column is found."
+  [a-query stage-number legacy-columns legacy-refs]
+  ;; Set up this query stage's `:aggregation` list as the context for [[convert/->pMBQL]] to convert legacy
+  ;; `[:aggregation 0]` refs into pMBQL `[:aggregation uuid]` refs.
+  (convert/with-aggregation-list (:aggregation (lib.util/query-stage a-query stage-number))
+    (let [columns       (mapv ->column-or-ref legacy-columns)
+          field-refs    (map legacy-ref->pMBQL legacy-refs)
+          ;; matches is a map of columns to the corresponding index in field-refs.
+          matches       (match-many field-refs
+                                    columns
+                                    [;; first try matching by IDs keeping the join information
+                                     find-closest-matches-for-refs-by-id-and-join
+                                     ;; then try a normal match
+                                     #(lib.equality/find-closest-matches-for-refs a-query stage-number %1 %2 {:keep-join? true})
+                                     ;; then try matching field refs having a name instead of an ID
+                                     #(find-closest-matches-for-refs-by-name a-query stage-number %1 %2)])
+          ;; We want to return a parallel list to field-refs, giving the index of the matching column (or -1).
+          ;; First, map each column to its index (in the column list).
+          column->index (into {} (for [index (range (count columns))]
+                                   [(nth columns index) index]))
+          ;; And use that to map each match's ref-index to its column-index.
+          by-index      (into {} (for [[column ref-index] matches]
+                                   [ref-index (column->index column)]))
+          res (->> (range (count legacy-refs))
+                   (map #(by-index % -1))
+                   to-array)]
+      res)))
+
```
I really don't like putting all this logic in `lib.js`, especially without writing any extra tests for it. It's a lot harder to test vs if it's in a `.cljc` file. Plus it's not available if other code needs to build on it in the rest of Metabase lib. 

Something like this should go in a regular `.cljc` namespace with actual tests, with wrappers around it only for doing conversion between JS and Cljs types

bshepherdson replied:
I agree with Cam. Most of this logic looks like it should already have been handled by `lib.equality/find-closest-matches-for-refs`.

metamben replied:
I does remove part of it, but not the whole. If we want to incorporate something like this in `lib.equality/find-closest-matches-for-refs`, then we'll have to make it optional too and probably disable it at all other call sites. Also note that the composition logic implemented here is a bit different. Here, we make sure that a match early in the series of tests cannot be overridden by a later match. I think `lib.equality/find-closest-matches-for-refs` just assumes there are no duplicate matches. (In fact, I've switched the order of arguments in a `merge` in that function to keep earlier matches, so there might not be a real difference anymore.)

bshepherdson commented:
I talked this over with @metamben and we have a plan worked out. I'm driving this issue again, now that I'm back.

Next steps are:
- Add some unit tests to cover some missing cases, to increase the safety of the next steps.
- Refactor `find-closest-matches-for-refs` and the extra logic in `metabase.lib.js` to be more unified.
    - Probably also making the "squinting strategy" pluggable, since in some cases we need different precise details of which columns are considered to match, and which aren't.

bshepherdson commented:
Another odd case:
- Create a saved query for Orders
- Create a new query based on the above.
- Add an implicitly joinable column, eg. `User -> State`
- It works correctly.
- Remove that `User -> State` column again.

Expected:
- `User -> State` disappears from the selected columns
-  And reappears in its original spot in the group of implicitly joinable columns on `Orders.USER_ID`.

Actual:
- It does disappear correctly.
- But it appears under a new `More columns` group separate from the implicit join for `USER_ID`.

**But** if I then re-run the query, it appears correctly. Is this intended? It looks like freshly hidden columns are handled specially, and once the query is re-run and fresh query metadata comes in, the column returns to normal. But I'm not sure if that special handling is deliberate, or naive and missing this case.

metamben commented:
> Is this intended?

I assumed this was an optimization in the FE so as not to re-run the query for every removed column.

bshepherdson committed:
[MLv2] Refactor lib.equality to "squinting"; integrate closest-matching-metadata

This fixes all the e2e and BE unit tests for viz settings #32501, but
with some caveats.

One e2e test is skipped - it's showing a real issue (see #32373) but
since it's a new test and not a regression, it can be tackled later.

The underlying issue is that we rely on `:lib/source` when generating refs
- For current stages with source-tables:
    - Use ID refs, `:expression` and `:aggregation` refs
- For previous stages and source-cards:
    - Use `:name` refs for everything.

Then for a nested query the columns have `:lib/source :source/card` at
first. But when you change the set of selected columns, a `:fields`
clause is added to the query, and all the columns change to
`:source/fields`.

`:source/card` refs use names, but `:source/fields` refs use IDs. This
breaks the `:selected?` logic and is causing the test failure.
I don't want to block this PR further while we rework `:lib/source` and
refs.

Also some would-be error cases in lib.equality matching actually occur
at least in tests. Removing the ambiguities proved to be difficult, and
perhaps some of these cases are not really errors? I'm planning to
rework the matching in a later PR anyway, so I've relaxed those errors
to logged warnings for now.

camsaul committed:
Fix failing test for #23293

camsaul commented on src/metabase/lib/js.cljs:
````diff
@@ -465,6 +469,82 @@
+  }
+  ```"
+  [a-query stage-number ^js js-options]
+  (let [^js js-options (or js-options #js {})
+        options        {:include-joined?              (gobj/get js-options "includeJoined" true)
+                        :include-expressions?         (gobj/get js-options "includeExpressions" true)
+                        :include-implicitly-joinable? (gobj/get js-options "includeImplicitlyJoinable" true)}
````
would this be easier if we just called `js->clj` on the JS object and then `update-keys` with `u/-:kebab-case-en` and passed that map in directly? Main reason I'm asking is because if we add a new option we'd get it for free that way and not need to hardcode it here as well

bshepherdson replied:
These options are now never passed as anything but `{}` by the FE, so I've just dropped the argument and all the conversion code.

camsaul commented on src/metabase/lib/js.cljs:
```diff
@@ -465,6 +469,82 @@
+    ;; It's already a :metadata/column map
+    column))
+
+(defn ^:export find-column-indexes-from-legacy-refs
+  "Given a list of columns (either JS `data.cols` or MLv2 `ColumnMetadata`) and a list of legacy refs, find each ref's
+  corresponding index into the list of columns.
+
+  Returns a parallel list to the refs, with the corresponding index, or -1 if no matching column is found."
+  [a-query stage-number legacy-columns legacy-refs]
+  ;; Set up this query stage's `:aggregation` list as the context for [[convert/->pMBQL]] to convert legacy
+  ;; `[:aggregation 0]` refs into pMBQL `[:aggregation uuid]` refs.
+  (convert/with-aggregation-list (:aggregation (lib.util/query-stage a-query stage-number))
+    (let [columns       (mapv ->column-or-ref legacy-columns)
+          field-refs    (map legacy-ref->pMBQL legacy-refs)
+          ;; matches is a map of columns to the corresponding index in field-refs.
+          matches       (lib.equality/find-closest-matches-for-refs a-query stage-number field-refs columns {:keep-join? true})
+          ;; We want to return a parallel list to field-refs, giving the index of the matching column (or -1).
+          ;; First, map each column to its index (in the column list).
+          column->index (into {} (for [index (range (count columns))]
+                                   [(nth columns index) index]))
+          ;; And use that to map each match's ref-index to its column-index.
+          by-index      (into {} (for [[column ref-index] matches]
+                                   [ref-index (column->index column)]))
+          res (->> (range (count legacy-refs))
+                   (map #(by-index % -1))
+                   to-array)]
+      res)))
```
Functions like this should go in the `cljc` files so they're easy to work with and test from a Clojure REPL, stuff in `js.cljs` should just be exports and light type conversion.

bshepherdson replied:
I'm not sure I agree.

Pushing this logic out to one place isn't so bad. But search for everywhere that `metabase.lib.convert/legacy-ref->pMBQL` is called - that shows dealing with legacy refs escaping from `metabase.lib.js` and finding its way into `lib.field`, `lib.filter`, and one or two more. That feels to me like something that should be dealt with at the edges, not throughout MLv2.

On the other hand, this is mixing JS conversion with some massaging of the return type of `find-closest-matches-for-refs`. I could see moving that inner part into `lib.equality` and adding it to the `discouraged-vars` linter so no one else calls it. That would leave just the JS (and `legacy-ref->pMBQL`) conversions here.

bshepherdson replied:
I've moved the guts of this to `lib.equality` and set that function as a `:discouraged-var`.

camsaul commented on src/metabase/lib/metadata/calculation.cljc:
```diff
@@ -92,6 +94,9 @@
                          :stage-number stage-number}
                         e)))))))
 
+(defmethod display-name-method :dispatch-type/nil [_query _stage-number _x _stage]
+  "")
```
This seems like kind of an icky hack. Shouldn't the display name of `nil` be `nil`?

bshepherdson replied:
Removed.

camsaul commented on test/metabase/lib/test_util.cljc:
```diff
@@ -264,6 +280,12 @@
                                        (mapv #(dissoc % :id :table-id)))}))])))
         (meta/tables)))
 
+(defn metadata-provider-with-mock-card [card]
+  (lib/composed-metadata-provider
+    meta/metadata-provider
+    (mock-metadata-provider
+      {:cards [card]})))
```
I wrote almost this exact same utility in 

https://github.com/metabase/metabase/pull/33429/files#diff-1da1a58aabf037690d9d1a71c9254ccfdb75eecb7b39f1888f3644ed0d37441aR10-R26

bshepherdson replied:
There's a lot of pieces of this branch that will need consolidating. Most of the `lib.js` functionality is partially duplicated from filters work, etc. I'll clean it up in a following PR.

bshepherdson commented:
I've just pushed a new version with 6 squashed, presentable commits in place of the original 57(!) commits. I've still got a side branch with the originals.

It's still broken out by author, and while I'm not sure the tests will pass at each commit the overall flow of the changes makes sense. It's possible there's some auto-merge damage in there somewhere, so we'll see how the CI run goes.

metamben commented on src/metabase/lib/equality.cljc:
```diff
@@ -104,106 +109,365 @@
-  like [[metabase.lib.breakout/breakoutable-columns]] which are supposed to include `:breakout-position` for columns
-  that are already present as a breakout; sometimes the column in the breakout does not exactly match what MLv2 would
-  have generated. So try to figure out which column it is referring to.
+(mu/defn resolve-field-id :- lib.metadata/ColumnMetadata
```
I wouldn't be able to guess this is the namespace to look for this function. I would look for it in some metadata namespace.

bshepherdson replied:
Yeah, I agree that this location sucks (see the docstring).

It used to be in `lib/field` and only used from there. Now it's used from `lib/field` and `lib/equality`. `lib/field` requires `lib/equality`. `resolve-field-id` requires `lib.card/saved-question-metadata`, and of course everyone, including `lib.card`, requires `lib.metadata.calculation`.

So I can see these options:
1. Move `resolve-field-id`, `lib.card/saved-question-metadata`, and the latter's private helpers to somewhere shared, probably `lib.metadata.calculation`.
2. Move `resolve-field-id` to a new namespace, perhaps `metabase.lib.metadata.resolve` (`resolvers`?)
3. Leave it where it is.

If we're unhappy with where it is in this PR then I think option 2 is the way to go. Thoughts?

metamben commented on src/metabase/lib/equality.cljc:
```diff
@@ -104,106 +109,365 @@
+            (nil? ref-fn))            result
+        (nil? ref-fn)                 (recur current-needles current-haystack more-ref-fns result)
+        :else
+        (let [haystack-ref->col  (m/index-by ref-fn current-haystack)
```
Should it be an error if `(< (count haystack-ref->col) (count current-haystack))`?

metamben commented on src/metabase/lib/equality.cljc:
```diff
@@ -104,106 +109,365 @@
+  - If both inputs are nonempty, continue to the next `ref-fns`, if any."
+  [needles haystack-metadatas opts ref-fns]
+  (let [original-needles (annotate-index needles)]
+    (loop [current-needles         original-needles
```
What about keeping `current-needles` as vectors to make lookup by position faster?

metamben commented on src/metabase/lib/equality.cljc:
```diff
@@ -104,106 +109,365 @@
+        (or (empty? current-needles)
+            (empty? current-haystack)
+            (nil? ref-fn))            result
+        (nil? ref-fn)                 (recur current-needles current-haystack more-ref-fns result)
```
The previous line must have matched before this matches. It would be an infinite loop anyway.

metamben commented on src/metabase/lib/equality.cljc:
```diff
@@ -104,106 +109,365 @@
+(defn find-closest-matches-for-refs
+  "For each of `needles` (which must be a ref) find the column or ref in `haystack` that it most closely matches. This
+  is meant to power things like [[metabase.lib.breakout/breakoutable-columns]] which are supposed to include
+  `:breakout-position` for columns that are already present as a breakout; sometimes the column in the breakout does not
```
Isn't this a bad example? Is this function recommended when matching against columns?

metamben commented on src/metabase/lib/equality.cljc:
```diff
@@ -104,106 +109,365 @@
+     (cascading-matches
+       needles haystack-metadatas opts
+       [;; The first pass attempts to match the refs using the regular [[find-closest-matches-for-refs]].
+        lib.ref/ref
```
I still have the feeling that this can result in a worse match than optimal by ignoring a good match by ID of a column from a source card (generating a ref by name) in favor of a worse match from a table (generating a ref by ID).

metamben commented on src/metabase/lib/equality.cljc:
```diff
@@ -104,106 +109,365 @@
+                     haystack-metadatas
+                     opts))]
+     ;; Prefer the more precise integer ID matches over the name-based ones.
+     (merge by-name basic))))
```
Would you pass in `query` whenever you can? Or are there situations when this could lead to undesired matches?

I have the feeling we have now lots of options and now rules which one to use. I'm afraid we'll be just experimenting which call makes the tests pass...

If you have some good rules, we should put them in the namespace docstring and/or add an explanation to each of these functions about the correct use. (I know there is already some documentation, but I'm not sure it's (precise) enough.)
